### PR TITLE
feat: added find subsets algorithm using bitmanipulation

### PR DESCRIPTION
**Pull request by Nikunj-bisht**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [X] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [X] This pull request is all my own work -- I have not plagiarized.
- [X] I know that pull requests will not be merged if they fail the automated tests.
- [X] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [X] All new JavaScript files are placed inside an existing directory.
- [X] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [ ] All new algorithms have a URL in their comments that points to Wikipedia or another similar explanation.
- [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Bit-Manipulation/GenerateSubSets.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * @example [1,2] -> [[],[1],[2],[1,2]]
+ */
+
+// The time complexity of this algorithm is BigO(2^n) where n is the length of array
+function generateSubSets(inputArray) {
+  if (!Array.isArray(inputArray)) {
+    return 'Input is not an array'
+  }
+  let arrayLength = inputArray.length
+  let subSets = []
```
Comment (appgurueu) [inline Bit-Manipulation/GenerateSubSets.js:14]: Could use `Array(1 << arrayLength)` to allocate this to the right size from the get go, then you could directly use `subSets[i] = subSet` below.

File: Bit-Manipulation/GenerateSubSets.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * @function generateSubSets
+ * @param {Array} inputArray
+ * @returns {Array}
+ * @example [1,2] -> [[],[1],[2],[1,2]]
+ */
+
+// The time complexity of this algorithm is BigO(2^n) where n is the length of array
+function generateSubSets(inputArray) {
+  if (!Array.isArray(inputArray)) {
+    return 'Input is not an array'
```
Comment (appgurueu) [inline Bit-Manipulation/GenerateSubSets.js:11]: This should throw, not `return`.

**Review by appgurueu — COMMENTED** [review]
Should also have an error case if the array is too large (more than 32 elements, since you only have 32 bits). This error case should be tested.

Otherwise fine!

**Comment by Nikunj-bisht** [comment]
@appgurueu made the required changes

File: Bit-Manipulation/test/GenerateSubSets.test.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { expect } from 'vitest'
```
Comment (appgurueu) [inline Bit-Manipulation/test/GenerateSubSets.test.js:1]: I don't think you need to import `expect`. You should also test the error cases (at least the `RangeError`) btw.

**Comment by Nikunj-bisht** [comment]
@appgurueu done with all the changes

File: Bit-Manipulation/GenerateSubSets.js [hunk 4]
```diff
@@ -8,7 +8,7 @@
 // The time complexity of this algorithm is BigO(2^n) where n is the length of array
 function generateSubSets(inputArray) {
   if (!Array.isArray(inputArray)) {
-    throw new TypeError('Provided input is not an array')
```
Comment (appgurueu) [inline Bit-Manipulation/GenerateSubSets.js]: Huh? What's the issue with throwing a `TypeError`? It seems appropriate here.

**Review by appgurueu — APPROVED** [review]

**Comment by Nikunj-bisht** [comment]
@appgurueu @raklaptudirm can we merge this pr?

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**